### PR TITLE
Fix password reset flow

### PR DESCRIPTION
**Pull request by tijmenb**
The password reset flow is currently broken. It should work like this:

1. When users click on a time-expired reset password link, they should see an error message
2. When users click on a previously used or invalid reset password link, they should see an error message
3. When users provide an previously used password, too-weak password or two different passwords, they should see validation errors

Currently there is no validation on the token validity or expiry when the password reset page is loaded, and when the user provides an invalid input, they will see a generic, possibly incorrect error page instead of the password form with errors.

This means users can click on an expired password link and think everything is okay, input a new password and see a message that their link is invalid. Also, users can input an incorrect password and get the message "It may be that you’ve used this passphrase already, or the link you were using has expired or was invalid". We've had several Zendesk requests from people who couldn't change their passwords.

For example: https://govuk.zendesk.com/agent/tickets/1176278

This commit fixes the flow, and updates the integration tests to cover the scenarios.

This commit does not introduce new behaviour, this is the original flow before a change made related to the big Devise upgrade of this summer (930c07c for example).

## Before

screen shot 2015-11-03 at 10 27 01

Visiting the page with an incorrect token.

screen shot 2015-11-03 at 10 26 53

Typing in a too weak password.

## After

screen shot 2015-11-03 at 10 22 45

Visiting the page with an incorrect token.

screen shot 2015-11-03 at 10 23 39

Typing in a too-weak password.

/cc @fofr

### CODE REVIEW

**Comment by boffbowsh** [comment]
Looks good to me from a code POV. Has copy been checked?

**Comment by fofr** [comment]
It's just the one line of copy that's changed as far as I can see, and that clarifies the situation in which the "Try again" error message appears.

File: app/controllers/passwords_controller.rb [hunk 1]
```diff
@@ -2,6 +2,16 @@ class PasswordsController < Devise::PasswordsController
   before_filter :record_password_reset_request, only: :create
   before_filter :record_reset_page_loaded, only: :edit
 
+  def edit
+    super
+
+    user = user_from_params
+    unless user && user.reset_password_period_valid?
+      render 'devise/passwords/reset_error'
+      return
```
Comment (benlovell) [inline app/controllers/passwords_controller.rb]: Is this necessary?

File: app/controllers/passwords_controller.rb [hunk 2]
```diff
@@ -16,13 +26,12 @@ def update
...
         record_password_reset_failure(resource) if resource.persisted?
-        render 'devise/passwords/reset_error'
-        return
       end
     end
   end
 
-  protected
+protected
+
```
Comment (benlovell) [inline app/controllers/passwords_controller.rb]: Style guide says we should remove this line of whitespace.

**Comment by tijmenb** [comment]
I've updated this PR.

**Comment by benlovell** [comment]
Code looks good. :ship:

**Comment by jamiecobbett** [comment]
This looks awesome, thanks @tijmenb

**Pull request merged**